### DISCUSSION TITLE
All states and events should have an attribute that shows who (or what) triggered it

### DISCUSSION THREAD

**Discussion by schwickster** [home-assistant/architecture#722] [Entity Models] [answered]
Hey,

Currently it's impossible to know who or what triggered a state or event (from a state or event perspective). IE: Did my wife press the light button in the companion app? Did I? Or maybe one of the kids pressed the light switch. Sure you can check the logbook to see who did it. But from an automation perspective it's not possible because the logbook isn't updated when the state or event is triggered.

I propose that each state and event should show which user, device or automation triggered the state. If a user used the front-end to turn on an entity, than said user is shown. If an automation is triggered, by default Home Assistant will be shown or can manually be set by the person creating the automation. Last but not least, if a device triggers something (IE: A light switch which controls the lights) than said device will be shown.

What will this accomplish? A couple of things:
- Debugging to see who or what triggered something
- Preference automation (If wife pressed it, lights need to be a certain colour or brightness, if I pressed, just normal lights, if pressed using button, just act normal)
- Make choices after an event/state is triggered depending on who or what did it. IE: If a sensor turned on the lights, do nothing. If I turned on the lights on manually, than I need to turn off the motion sensor to prevent the light from going off. Currently it requires additional logic to keep track of who did what. It's a mess to accomplish this.
- Personalized announcement using smart speakers or messages. IE: Goodmorning PERSONWHOTRIGGERDSOMETHING or PERSON turned on the alarm

In the event or state next beside last_updated a property needs to be added called last_updated_by. Said property can be also be manually set to assign a custom value.

Hopefully you can all agree this is a good idea and in the future it will be implemented.

**Comment by frenck** [comment]
What would need to change and how do you plan to approach this?
Generic implementation details of your proposal are missing at this point.

**Reply by schwickster** [reply]
As mentioned in my post (If I understand your question correctly) is to add an additional property called last_updated_by which contains a string. This string can either be manually set in automations or integrations (IE: Nodered) or automatically set (if no manual overwrite is given) by HA fetching the user who triggered in. In case no user has triggered it or no manual overwrite, default to HA.

**Reply by frenck** [reply]
But isn't context already available for this?
The context currently contains the user and the parent context.

**Reply by frenck** [reply]
Please forgive my ignorance, but is this a feature request? Or are you working on an implementation?

**Reply by schwickster** [reply]
Yes it's a feature request. But I thought I wanted to propose it here because it requires an architecture change first so that every integration and core things can add this idea. My apologies if this wasn't the best place to propose this architecture change. 

I would love to implement this.. but at the moment it's a daunting task for me. I got some other ideas I maybe want to implement before I start a daunting thing like this... I might look into it.. but doubt I'll succeed.

To answer your other question: As far as I can see in nodered when I change the state of a light I can't see who triggered the change. I can see when it was changed and what changed, but not who (or what). As mentioned I use custom logic in nodered to keep track of this, but it would be a better idea to see it as a core. Some other integrations I can see this property is present, but not filled.

**Reply by frenck** [reply]
> Yes it's a feature request. But I thought I wanted to propose it here because it requires an architecture change first so that every integration and core things can add this idea. My apologies if this wasn't the best place to propose this architecture change.

> I would love to implement this.. but at the moment it's a daunting task for me.

For feature requests, please use our community forum instead:

<https://community.home-assistant.io/c/feature-requests/13>

We should not be discussing and wasting time on architectural proposals that have no intention of being implemented upfront.

> As far as I can see in nodered when I change the state of a light I can't see who triggered the change.

That is a Node-RED issue and feature request I would say.

>  but it would be a better idea to see it as a core.

It already is in the core.

![image](https://user-images.githubusercontent.com/195327/152133696-1998627b-3bbe-41ba-a2de-1c79eaa83094.png)

**Comment by frenck** [comment, marked as answer]
Marking this discussion as resolved, as it is a feature request.
